Spigot-Server-Patches/0747-Add-feature-access-to-custom-ChunkGenerators.patch

```diff
@@ -0,0 +1,171 @@
++        if(updateLiquids && ((CraftBlockData) data).getState().getBlock() instanceof BlockFluids) {
++            getDelegate().getFluidTickList().queue(position, ((BlockFluids) ((CraftBlockData) data).getState().getBlock()).getFluidState(((CraftBlockData) data).getState()).getType(), 0);
++        }
++    }
++
++    @Override
++    @SuppressWarnings("deprecation")
++    public @NotNull World getWorld() {
++        // getMinecraftWorld is deprecated in RLWA, because reading/writing the returned world causes a deadlock.
++        // By implementing this, and covering it in warnings, we're assuming people won't be stupid, and
++        // if they are stupid, they'll figure it out pretty fast.
++        return getDelegate().getMinecraftWorld().getWorld();
++    }
++
++    @Override
++    public @NotNull BlockData getBlockData(int x, int y, int z) {
++        return CraftBlockData.fromData(getDelegate().getType(new BlockPosition(x, y, z)));
++    }
++
++    @Override
++    public int getCenterChunkX() {
++        return getDelegate().getCenterX();
++    }
++
++    @Override
++    public int getCenterChunkZ() {
++        return getDelegate().getCenterZ();
++    }
++
++    @SuppressWarnings("unchecked")
++    @Override
++    public <T extends Entity> @NotNull T spawn(@NotNull Vector location, @NotNull Class<T> clazz, @Nullable Consumer<T> function, CreatureSpawnEvent.@NotNull SpawnReason reason) throws IllegalArgumentException {
++        net.minecraft.world.entity.Entity entity = getDelegate().getMinecraftWorld().getWorld().createEntity(location.toLocation(getWorld()), clazz);
++        Preconditions.checkArgument(entity != null, "Cannot spawn null entity");
++
++        if (entity instanceof EntityInsentient) {
++            ((EntityInsentient) entity).prepare(getDelegate(), getDelegate().getDamageScaler(entity.getChunkCoordinates()), EnumMobSpawn.COMMAND, null, null);
++        }
++
++        if (function != null) {
++            function.accept((T) entity.getBukkitEntity());
++        }
++
++        getDelegate().addEntity(entity, reason);
++        return (T) entity.getBukkitEntity();
++    }
++
++    @NotNull
++    private RegionLimitedWorldAccess getDelegate() {
++        return Objects.requireNonNull(world.get(), "Cannot access ProtoWorld after generation!");
++    }
++}
+diff --git a/src/main/java/net/minecraft/server/level/RegionLimitedWorldAccess.java b/src/main/java/net/minecraft/server/level/RegionLimitedWorldAccess.java
+index 04006caeeb42b523d986efc313828557854718d7..2db5377615313100481df8fc64305d15abd64f79 100644
+--- a/src/main/java/net/minecraft/server/level/RegionLimitedWorldAccess.java
++++ b/src/main/java/net/minecraft/server/level/RegionLimitedWorldAccess.java
+@@ -93,10 +93,14 @@ public class RegionLimitedWorldAccess implements GeneratorAccessSeed {
+         }
+     }
+ 
++    public int getCenterX() { return a(); } // Paper - obfhelper
++
```

Comment:
These blank lines between obfhelpers and the method they are for should be removed.